Step3 자동차 경주

janghomoon opened the pull request:
안녕하세요 
이전 말씀하신 부분이랑 보완해서 했는데 잘되었는지 모르곘습니다. 
검토해주시면 감사합니다.

jhmoon-howser committed:
feat: 입력 부 구현 완료,(todo리스트 작성)

jhmoon-howser committed:
feat: 레이스 및 , 자동차 엔티티 생성

jhmoon-howser committed:
feat: 매 개임 자동차별 0-9 사이 수중 4보다 클 경우 이동 수 1을 반환하는 기능 구현

jhmoon-howser committed:
feat: 게임 실행  및 출력 부분 구현 완료

catsbi commented on src/main/kotlin/study/racingcar/Application.kt:
```diff
@@ -0,0 +1,17 @@
+package study.racingcar
+
+class Application {
+    fun main() {
+        println("자동차 대수는 몇 대인가요?")
+        var carCount:Int? = readln().toIntOrNull()
+        inputValid(carCount)
+        print("시도할 횟수는 몇 회인가요?")
+        val moveCount:Int? = readln().toIntOrNull()
+        inputValid(moveCount)
+    }
```
장호님 현재 부트스트랩 클래스가 없다보니 게임을 할 수 없는상황으로 보입니다. 즉 미완성 프로그램으로 보이는데 확인 부탁드려요. 

추가적으로 해당 코드 피드백을 드리자면, 이 클래스는 입력을 받은 뒤 무슨 행동을 하는 함수인가요? 사실상 값을 받고 유효성 체크 이후 아무것도 하지 않는 기능으로 보입니다 🤔 

기본적인 MVC pattern을 기반으로 프로젝트 구조를 고민해보면 좋을 것 같아요. 
참고: https://velog.io/@langoustine/%EC%97%AC%EA%B8%B0%EB%8F%84-MVC-%EC%A0%80%EA%B8%B0%EB%8F%84-MVC-MVC-%ED%8C%A8%ED%84%B4%EC%9D%B4-%EB%AD%90%EC%95%BC

catsbi commented on src/main/kotlin/study/racingcar/Application.kt:
```diff
@@ -0,0 +1,17 @@
+        inputValid(moveCount)
+    }
+
+    private fun inputValid(inputCount:Int?) {
+        if (inputCount == null) throw IllegalArgumentException("숫자를 입력해주세요.")
+        require(inputCount.toInt() > 0) { throw IllegalArgumentException("1 이상의 값을 입력해주세요.") }
+    }
```
if문은 block이 생략되어있고 require 에는 block `{}`이 작성되어 있습니다. 

그 차이는 뭘까요? 
어디에는 블럭`{}` 이 생략되고 어디에는 있고 이렇게 일관성이 떨어질 경우 코드의 가독성은 저해될 수 밖에 없습니다. 
그래서 저는 IDE 자동생성되는 코드를 포함하여 블럭 생략을 지양하라고 피드백을 항상드리는 편입니다. 

장호님은 블럭 생략 기준이 있으실까요?

catsbi commented on src/main/kotlin/study/racingcar/entity/Car.kt:
```diff
@@ -0,0 +1,11 @@
+package study.racingcar.entity
+
+class Car(var moveCount: Int = 0) {
```
moveCount가 var이기에 외부에서 위변조가 너무 쉽게 될 것 같습니다. 
java로 치면 public int moveCount; 인것이죠. 

어떻게 외부의 위변조를 막을 수 있을까요?

catsbi commented on src/main/kotlin/study/racingcar/entity/Car.kt:
```diff
@@ -0,0 +1,11 @@
+package study.racingcar.entity
+
+class Car(var moveCount: Int = 0) {
+    fun move(distance: Int) {
+        moveCount += distance
+    }
```
객체를 보면 move가 단순히 setter역할만을 수행하고 있습니다. 
이래선 Car가 도메인이라기보단 VO 객체이지 않을까 생각이 드는데요. 적절한 책임일까요?

catsbi commented on src/main/kotlin/study/racingcar/entity/Car.kt:
```diff
@@ -0,0 +1,11 @@
+        moveCount += distance
+    }
+
+    fun getResult(): String {
+        return "-".repeat(moveCount)
+    }
```
만약, 기획이 바뀌거나 다른 요구사항으로 스키드마크가 "-"가 아니라 "="로 바뀌어야 한다면 어떨까요? 
아니면 도로 노선이나 시간에 따라 언제는 "-" 언제는 "=" 가변적으로 바뀌어야 한다면 또 어떨까요? 이 코드를 매 번 바꿔줘야 할까요?

도메인 계층에서 출력(UI) 계층에 대한 책임을 지고 있다보니 단일 책임 원칙(SRP)도 위배되는 것 같네요. 출력에 대한 책임은 출력 계층에게 위임하는게 어떨까요?

catsbi commented on src/main/kotlin/study/racingcar/entity/Race.kt:
```diff
@@ -0,0 +1,3 @@
+package study.racingcar.entity
+
+class Race(val carCount: Int, val raceCount: Int)
```
단순 값 저장 객체라면 data class를 고려해보는건 어떨까요?

catsbi commented on src/main/kotlin/study/racingcar/service/RacingService.kt:
```diff
@@ -0,0 +1,25 @@
+
+import study.racingcar.entity.Car
+
+class RacingService(val carCount: Int, val raceCount: Int) {
```
Service 계층에서는 별도의 상태를 가지는것을 권장하지 않습니다. 
그 이유가 뭘까요?

catsbi commented on src/main/kotlin/study/racingcar/service/RacingService.kt:
```diff
@@ -0,0 +1,25 @@
+        }
+    }
+    
+    fun getRandomMoveCount(): Int {
+        val randDomMoveCount = (0..9).random()
+        return if (randDomMoveCount > 4) 1 else 0
+    }
```
자동차마다 속력이 달라서 어떤 자동차는 1이 아니라 2씩 전진해야한다면 어떨까요? 

그리고 값이 random이 아닌 전략에 의해 정해지고 0~9가 아니라 -5~10이라면 어떨까요? 
객체지향 프로그래밍에서는 SOLID 원칙을 지키는 것을 권장하는데요. 이 로직은 OCP를 위배할 것 같아요. 

전략패턴(Strategy Pattern)등을 활용해서 추상화해볼 수 없을까요?

catsbi commented on src/main/kotlin/study/racingcar/service/RacingService.kt:
```diff
@@ -0,0 +1,25 @@
+        }
+    }
+
+    fun result() {
+        cars.forEach { e ->
+            println(e.getResult())
+        }
+    }
```
입/출력에 대한 책임을 모두 출력 계층으로 분리해보는건 어떨까요?

catsbi reviewed (changes requested):
장호님 3단계 자동차 경주 미션을 잘 시작해주셨어요 👍 

그런데 현재 프로그램이 시작할 수 있는 부트스트랩 클래스가 보이지 않고, 임의로 생성해서 실행시켜보니 게임을 진행할수 없었습니다. 요구사항이 제대로 구현되지 않은것 같아 피드백과 함께 참고해서 코드 추가 이후 다시 리뷰요청 부탁드려요.

jhmoon-howser committed:
fix: Application 함수 변경

입력이후 컨트롤러를 호출하도록 변경하였고  호출 된 컨트롤러에서 해당 파라메터 유효성 검사를 하도록 수정하였습니다.

jhmoon-howser committed:
fix: race 서비스 수정 및 도메인 수정

피드백으로 주셨던 부분 들 수정 해봤습니다.

코틀린을 이번에 처음 사용하다보니 확인하고 작성하고 있는데

혹시 'Service 계층에서는 별도의 상태를 가지는것을 권장하지 않습니다'

이부분 수정을 어뗗가하면 좋을지 알 수 있을까요?

'입/출력에 대한 책임을 모두 출력 계층으로 분리해보는건 어떨까요?'

입력값 -> controller 호출 받고 결과값 까지 모두 찍었는데 출력 계층이라 함은  파라메터 최초 읽는 Application 파일을 말씀하시는걸까요?

catsbi commented on src/main/kotlin/study/racingcar/Application.kt:
```diff
@@ -1,17 +1,13 @@
-    }
+import study.racingcar.controller.RacingCarController
+
+fun main() {
+    println("자동차 대수는 몇 대인가요?")
+    val carCount:Int? = readln().toIntOrNull()
+    print("시도할 횟수는 몇 회인가요?")
+    val gameCount:Int? = readln().toIntOrNull()
+    val racingCarController = RacingCarController()
+    racingCarController.start(carCount, gameCount)
```
UI에 대한 책임도 InputView나 OutputView같은 계층이 나와서 책임을 위임시켜보면 어떨까요? 
이펙티브 자바의 내용중에서 추상화의 레벨을 동등하게 유지하라는 항목이 있는데요. 

한 블럭에서 어떤 기능은 직접 연산을 수행하고 어떤 기능은 함수호출을 통해 진행되는 코드는 가독성이 떨어질 수 밖에 없거든요. 
여기선 입/출력에 대한 책임은 직접 수행하는데, 게임 진행은 컨트롤러에게 하고, 근데 출력도 컨트롤러가 책임지고 있습니다. 

UI와 Logic이 뒤섞여있죠. 이른 부분들을 맞춰주는게 가독성에 중요한 요인이 됩니다.

catsbi commented on src/main/kotlin/study/racingcar/controller/RacingCarController.kt:
```diff
@@ -0,0 +1,20 @@
+        race.printResult()
+    }
+
+    private fun inputValid(inputCount: Int?): Int {
+        if (inputCount == null) {
+            throw IllegalArgumentException("숫자를 입력해주세요.")
+        }
+        require(inputCount.toInt() > 0) { throw IllegalArgumentException("1 이상의 값을 입력해주세요.") }
+        return inputCount
+    }
```
입출력에 대한 책임을 컨트롤러에서 책임져야할까요?

catsbi commented on src/main/kotlin/study/racingcar/controller/RacingCarController.kt:
```diff
@@ -0,0 +1,20 @@
+    fun start(carCount: Int?, gameCount: Int?) {
+       val race = RacingService(inputValid(carCount), inputValid(gameCount))
+        race.start()
+        race.printResult()
```
input은 application에서 이뤄져서 CLI 을 application에서 책임지는줄알았는데 결과 출력은 controller에서 이뤄지고 있네요 🤔

catsbi commented on src/main/kotlin/study/racingcar/entity/Car.kt:
```diff
@@ -1,11 +1,14 @@
+
     fun move(distance: Int) {
-        moveCount += distance
+        if (distance > 3) moveCount ++
```
매개변수가 이름이 distance인데 3이상일 때만 count가 증가한다는게 뭔가 좀 이상하지 않나요. 
이름만 가지고 분석해보면 해당 자동차는 거리가 4 이상일 경우에만 움직임 횟수를 증가한다. 라고 보여요. 

매개변수명과 인스턴스변수명을 좀 더 시그니처를 보여줄 수 있게 네이밍하는게 중요해보입니다. 

그리고 해당 자동차가 3이 아닌 4 이상일 때 전진하고싶다면 이 코드를 수정해야할까요? 
그럼 그 다음 오전에는 3 이상일 경우, 오후에는 5 이상일 경우 전진하도록 하고싶다면 이 코드는 어떻게 수정해야 할까요? 

이렇게 기획이 바뀔때마다 코드가 수정되야한다면 SOLID원칙 중 OCP개방 폐쇄 원칙을 위배할 것 같은데요. 

디자인 패턴 중 전략 패턴을 활용해보는건 어떨까요? 
행위를 추상화하여 구현체에 의존하지 않도록 하는 것이죠.

참고: https://catsbi.oopy.io/344dbe7b-9774-48fc-9c95-b554e9c1c4bc

catsbi commented on src/main/kotlin/study/racingcar/entity/Car.kt:
```diff
@@ -1,11 +1,14 @@
 
-    fun getResult(): String {
-        return "-".repeat(moveCount)
+    fun getResult(symbol:String): String {
+        return symbol.repeat(moveCount)
```
도메인은 상태만 책임지도록 하는걸 권장드려요 출력은 UI에서 책임지도록 하는것이죠. 그럼 변환의 주체가 도메인이 아닌 UI로 바뀌게 됩니다. 

```suggestion
    fun getMoveCount(): Int = moveCount
    
    ...
class OutputView {
    ...
    
    fun view(racingHistories: List<RacingHistory>) {
        racingHistories.joinToString {
           val count = history.getMoveCount()
           
           SKID_MARK.repeat(count)
        }
    }
}
```

이런식으로 (예시용 수도 코드입니다) 도메인은 상태값만을 제공하고, 이를 가공하는건 UI에서 책임지는것이죠.

catsbi commented on src/main/kotlin/study/racingcar/service/RacingService.kt:
```diff
@@ -1,25 +1,28 @@
-        val randDomMoveCount = (0..9).random()
-        return if (randDomMoveCount > 4) 1 else 0
+
+    private fun getRandomMoveCount(): Int {
+        return (0..9).random()
+    }
```
4 이상일 경우 전진한다는 조건은 달라졌지만, 0~9까지의 랜덤값에 의해 결정된다는 구체화 정보는 그대로 있습니다. 이 역시 추상화가 가능할 것 같네요!

catsbi commented on src/main/kotlin/study/racingcar/service/RacingService.kt:
```diff
@@ -1,25 +1,28 @@
+import study.racingcar.entity.Car
 
-    val cars = List(carCount) { Car() }
+class RacingService(carCount: Int, private val gameCount: Int) {
```
서비스 계층은 무상태(stateless)를 지향합니다. 

여러번 재사용이 가능하도록 말이죠. 지금처럼 상태를 가지고 있을 경우, 레이싱을 할 때마다 새로 레이싱 서비스를 만들어야 하지 않을까요? 굳이 그러지 않아도 되는 계층인데 말이죠.

catsbi commented on src/main/kotlin/study/racingcar/service/RacingService.kt:
```diff
@@ -1,25 +1,28 @@
-    fun result() {
-        cars.forEach { e ->
-            println(e.getResult())
+    fun printResult() {
```
왜 서비스에서 출력을 책임져야 할까요?

catsbi commented on src/test/kotlin/study/racingcar/entity/CarTest.kt:
```diff
@@ -1,20 +1,19 @@
 package study.racingcar.entity
```
UI 계층(ex: controller, input, output)을 제외한 모든 객체는 테스트 코드가 작성되야 합니다. 
TDD라면 "테스트부터" 작성되야 하지만, 처음엔 쉽지 않기 때문에 이런 정통적인 bottom-up방식이 아닌 top-down방식도 괜찮습니다. 

테스트 케이스를 보충해볼까요?

catsbi reviewed (changes requested):
장호님 피드백 반영하느라 고생하셨습니다. 👍 

그런데 아직 좀 더 고민해봐야할 부분들이 있어서 코멘트 남겨드렸으니 확인 후 다시 리뷰요청 부탁드려요!

jhmoon-howser committed:
fix: 피드백 반영을 위한  코드 재수정

이번 수정에는 최대한 코드 를 변환하는 데 집중해서 수정했습니다.

피드백을 보고 수정을 다시 잔행하는 과정에서  컨트롤러에서 input valid 하는 부분은 제생각에는  큰 문제가 없는게 아닌지 궁금합니다.

또 input, output 즉 view 역할로 하는것도 말씀하신것을 듣고 이해했지만 outputview? 에서  '-' 표시를 하는것은 잘 이해가 가지 않습니다.
예를들어 자동차 리스트를 화면에 내려준다면  응답객체를 만들어 줄때 변환해서 내려준다던지..

매번 꼼꼼히 검토해주셔서 감사합니다.

catsbi commented on src/main/kotlin/study/racingcar/configuration/RacingConfiguration.kt:
```diff
@@ -0,0 +1,19 @@
+import study.racingcar.repository.CarRepository
+import study.racingcar.service.RacingService
+
+class RacingConfiguration {
```
설정 계층을 추가하셨네요 👍 
설정계층에서 구현체에 대한 의존성을 중앙 관리할 경우 실제 프로덕션 영역에서는 DIP에 좀 더 집중할 수 있게 되죠.

catsbi commented on src/main/kotlin/study/racingcar/entity/Car.kt:
```diff
@@ -4,11 +4,10 @@ class Car(private var moveCount: Int = 0) {
 
-    fun move(distance: Int) {
-        if (distance > 3) moveCount ++
+    fun move(
+        distance: Int,
+        validCondition: Int,
+    ) {
+        if (distance >= validCondition) moveCount++
```
결국 이 행위도 두 값의 비교로 행동이 결정된다는 점에서는 확장성이 떨어집니다. 다음과 같이 구현할 수 없을까요?
```suggestion
    fun moves(strategy: MoveStrategy) {
        if (strategy.movable()) {
            moveCount++
        }
```

catsbi commented on src/main/kotlin/study/racingcar/repository/CarRepository.kt:
```diff
@@ -0,0 +1,15 @@
+import study.racingcar.entity.Car
+
+class CarRepository {
+    private var cars: MutableList<Car> = mutableListOf()
```
var이면서 MutableList로 선언하신 이유가 있을까요?

catsbi commented on src/main/kotlin/study/racingcar/repository/CarRepository.kt:
```diff
@@ -0,0 +1,15 @@
+        this.cars.add(Car())
+    }
+    
+    fun findAllCars(): List<Car> {
+        return this.cars
+    }
```
외부에서 carRepository.findAllCars().clear() 
이래버리면 어떻게 될까요? 

외부 위변조에 취약할 것 같은데 어떻게 불변성을 높힐 수 있을까요?

catsbi commented on src/main/kotlin/study/racingcar/service/RacingService.kt:
```diff
@@ -1,28 +1,33 @@
+class RacingService(private val carRepository: CarRepository) {
+    fun intCars(carCount: Int) {
+        repeat(carCount) {
+            carRepository.saveDefaultCar()
```
레파지토리의 영역은 데이터를 스토어에 저장하는 것이지 도메인 생성의 책임은 없습니다. 
자동차를 생성하는 것 자체는 서비스에서 해줘도 될 것 같아요.

catsbi reviewed (approved):
장호님, 3단계 미션 피드백 반영 확인 했습니다 👍 

몇 가지 피드백 추가해드렸는데, 다음 미션 진행하며 같이 확인해보면 좋을 것 같아요. 
그리고 어려운부분이나 이해안가는부분 너무 혼자 고민하시면, 헤맬 수 있으니 언제든 편하게 DM 주세요. !
